### DISCUSSION TITLE
Find instances "almost" identical

### DISCUSSION THREAD

**Discussion by cadilhac** [ravendb/ravendb#19318] [Q&A]
Can RavenDb help on this matter:

In the db, I have a collection of objects representing official sets of settings. One object is:

settings/1:
{
    prop1: "off",
    prop2: "5",
    prop3: null,
    ....
    propN: "-4"
}

1. I am first interested in finding in the db an object that is exactly the same as the one in my query. This one is easy. I had talked about it in another post. I simply add a hash value to the object that represents the set of all the other values. I query by hash and it works well.
2. On the other hand, I would like to find objects that "look like" the one in the query. For instance, I could return to the user the object settings/33 because all its props except one for instance equal the props in the query. The response to the user would be "This set is almost what you are looking for". I would accept only a max of 2 differences. Does such a query exist? And if yes, does it mean that all of the properties of the Settings class must be indexed?

**Comment by ayende** [comment]
It looks like you are looking for the `More Like This` feature : https://ravendb.net/docs/article-page/6.2/Csharp/indexes/querying/morelikethis

**Comment by cadilhac** [comment]
Thank you. But it seems that the document to compare to has to be already in the db:

`.UsingDocument(x => x.Id == "articles/1")`

In my case, I want to supply a set of properties that are in memory, not in the db. Any way to do that?

**Reply by ayende** [reply]
Node that `UsingDocument` has an overload that accepts a JSON string - you can use that as the search target

**Reply by cadilhac** [reply]
Great. So now I use this:

```C#
var recipes = await RavenSession
.Query<FujiRecipe, FuzzyFujiRecipeIndex>()
.MoreLikeThis(builder => builder
.UsingDocument(jsonString))
.ToListAsync();
```

But I get a NullRef exception when one of the properties in the json string is null.

**Reply by cadilhac** [reply]
Anyway, I have temporarily removed the possible null values to avoid the exception, and I got a strange result. I passed a json string that is equivalent to one of the documents (I just changed one prop) so I was expecting to get one result. But instead I received 12, which is the whole set of my testing documents that are actually all very different.

**Reply by ayende** [reply]
I don't think that you can use a `null` there right now (may be a bug)
If you have a `null`, remove the property entirely.

**Reply by ayende** [reply]
Note that you are _querying_, and you'll get the documents in the order that they match.
The more they match, the more you'll get

**Reply by ayende** [reply]
The more they match, the "higher" they will be in the ranking. But you'll get those that aren't similar, because they are more similar than whatever else is there

**Reply by cadilhac** [reply]
> I don't think that you can use a `null` there right now (may be a bug) If you have a `null`, remove the property entirely.

I have to compare with the null values, so I can't remove these props. So maybe I can store a custom string in the index, like "NULL" and also have this in my json string for null props...

**Reply by cadilhac** [reply]
> Note that you are _querying_, and you'll get the documents in the order that they match. The more they match, the more you'll get

Understood.

**Reply by cadilhac** [reply]
> I don't think that you can use a `null` there right now (may be a bug) If you have a `null`, remove the property entirely.

Please, tell me if this is something you actually identify as a bug and if you intend to fix it in 5 and/or 6 branches.

**Comment by ayende** [comment]
Fixed here: https://github.com/ravendb/ravendb/pull/19884 - will go to 6.2
